### ISSUE TITLE
Form item ID: in which cases should the ID get created automatically?

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4999]
This popped up in another [issue](https://github.com/backdrop/backdrop-issues/issues/4997) and I wasn't able to figure out:

Are IDs for core form items supposed to always get generated automatically, or are there valid cases, where this shouldn't happen?

The downside of a missing ID is that the `for="the-id"` attribute can't get added, which means an accessibility regression.

**Comment by jenlampton** [comment]
> Are IDs for core form items supposed to always get generated automatically

yes. The idea is that there may be any number of forms on a page (news letter signup, comment form, search box in the header, search box in the footer, etc) and by allowing core to assign the IDs, then we can ensure that each one is unique. 

> The downside of a missing ID is that the for="the-input-id" attribute on the label can't get added, which means an accessibility regression.

Core should be adding these automatically, too. (I know it does for radios and checkboxes). If that's not happening, maybe we should look closer at why?

**Comment by klonos** [comment]
In #4497, we ended up manually adding the ID in that search box, to fix accessibility (see `admin_bar_links_search()`):

```php
function admin_bar_links_search() {
  $search['search'] = array(
    '#pre_render' => array(),
    '#attributes' => array(
      'class' => array('admin-bar-search'),
    ),
  );
  $search['search']['search'] = array(
    '#type' => 'search',
    '#title' => t('Admin search'),
    '#title_display' => 'invisible',
    '#attributes' => array(
      'placeholder' => t('Menu search'),
      'autocomplete' => 'off',
      'autocorrect' => 'off',
      'autocapitalize' => 'off',
    ),
    '#id' => 'admin-bar-search-items', // <-- why does this need to be added manually???
  );
  return $search;
}
```

**Comment by jenlampton** [comment]
This is great, we can do the same thing in admin bar :D

**Comment by klonos** [comment]
Sorry @jenlampton for messing the issue title. This was already done for the admin bar; we now need to ensure that this is consistent for all form items. `'#type' => 'search'` seems to not be doing it automatically.
